Test for proper serialization/deserialization of Hedera protobuf objects

deyanzz opened the pull request:
**Description**:
Introducing new tests to verify serialization & deserialization of Hedera protobuf objects.
Testing only methods `toProtobuf` and `fromProtobuf` in the following order:

---

**Related issue(s)**: [Issue #163](https://github.com/hashgraph/hedera-sdk-cpp/issues/163)

---

Fixes #
Added the following implementation files for each test suit

1. `ED25519PublicKeyTest.cc` - add test `ED25519PublicKeyToProtobuf`;
2. `ECDSAPublicKeyTest.cc` - add test `PublicKeyFromProtobuf`;
3. `ExchangeRateTest.cc` - add tests `DeserializeExchangeRateFromProtobuf` & 
`DeserializeExchangeRateSetFromProtobuf`;
4. `AccountBalanceTest.cc` - add test `DeserializeAccountBalanceFromProtobuf`;
5. `AccountIdTest.cc` - add tests `SerializeAccountIdToProtobuf` & `DeserializeAccountIdFromProtobuf`;
6. `ContractIdTest.cc` - add test `SerializeContractIdТoProtobuf` & `DeserializeContractIdFromProtobuf`;
7. `TransactionIdTest.cc` - add test `SerializeTransactionIdToProtobuf` & `DeserializeTransactionIdFromProtobuf`;
8. `TransferTest.cc` - add test `SerializeTransferToProtobuf` & `DeserializeTransferFromProtobuf`

Following classes already have tests for protobuf serialization / deserialization:
 - `TransactionReceipt`
 - `TransactionRecord`
 - `TransactionResponse`

The following classes couldn't be tested, because all of them seem to be internal:
 - class `Endpoint`
 - class `NodeAddress`
 - class `NodeAddressBook`
 - class `DurationConverter`
 - class `TimestampConverter`
---

**Notes for reviewer**:
These serialization / deserialization tests could be moved in the previously added test files to avoid adding new test suites named like `SerializationTests`, `DeserializationTests` and `SerializationAndDeserializationTests`.
For example:
 - tests `SerializeAccountIdToProtobuf` & `DeserializeAccountIdFromProtobuf` could be moved into `AccountId` as they are testing only class `AccountId`.
 - test `DeserializeTransactionRecordFromProtobuf` could be moved into `TransactionRecord` as they are testing only `TransactionRecord`.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

deyanzz commented on sdk/tests/SerializationAndDeserializationTests.cc:
```diff
@@ -0,0 +1,199 @@
+  EXPECT_EQ(protoAccountId->accountnum(), testAccountId.getAccountNum());
+}
+
+TEST_F(SerializationAndDeserializationTests, DeserializeAccountIdFromProtobufTest)
```
To write this test I used test [ProtobufAccountId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/AccountIdTest.cc#L125) from `AccountIdTest`.

deyanzz replied:
Test `SerializeAccountIdToProtobufTest` & `DeserializeAccountIdFromProtobufTest` moved into `AccountIdTest.cc` in commit [11ed0182be6d48d340a9897994290084e66cf20e].(https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/11ed0182be6d48d340a9897994290084e66cf20e)

deyanzz commented on sdk/tests/SerializationAndDeserializationTests.cc:
```diff
@@ -0,0 +1,199 @@
+  EXPECT_EQ(accountId.toString(), "123.456.789");
+}
+
+TEST_F(SerializationAndDeserializationTests, SerializeContractIdТoProtobufTest)
```
To write this test I used test [ProtobufContractId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/ContractIdTest.cc#L76) from ContractIdTest.

deyanzz replied:
Tests `SerializeContractIdТoProtobufTest` & `DeserializeContractIdFromProtobufTest` moved into `ContractIdTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/42c4b74b470cbb18745489aebed04b87c7b75235.

deyanzz commented on sdk/tests/SerializationAndDeserializationTests.cc:
```diff
@@ -0,0 +1,199 @@
+  EXPECT_EQ(protoContractId->contractnum(), testContractNum);
+}
+
+TEST_F(SerializationAndDeserializationTests, DeserializeContractIdFromProtobufTest)
```
To write this test I used test [ProtobufContractId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/ContractIdTest.cc#L76) from ContractIdTest.

deyanzz replied:
Test `DeserializeContractIdFromProtobufTest` moved into `ContractIdTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/42c4b74b470cbb18745489aebed04b87c7b75235.

deyanzz commented on sdk/tests/SerializationAndDeserializationTests.cc:
```diff
@@ -0,0 +1,199 @@
+  const std::chrono::system_clock::time_point mValidStart = std::chrono::system_clock::time_point();
+};
+
+TEST_F(SerializationAndDeserializationTests, SerializeAccountIdToProtobufTest)
```
To write this test I used test [ProtobufAccountId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/AccountIdTest.cc#L125) from `AccountIdTest`.

deyanzz replied:
Test `SerializeAccountIdToProtobufTest` moved into `AccountIdTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/11ed0182be6d48d340a9897994290084e66cf20e.

deyanzz commented on sdk/tests/SerializationAndDeserializationTests.cc:
```diff
@@ -0,0 +1,199 @@
+  EXPECT_EQ(contractId.getContractNum(), testContractNum);
+}
+
+TEST_F(SerializationAndDeserializationTests, SerializeTransactionIdToProtobufTest)
```
To write this test I used test [ProtobufTransactionId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransactionIdTest.cc#L51) from [TransactionIdTest](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransactionIdTest.cc).

deyanzz replied:
Test `SerializeTransactionIdToProtobufTest` moved into `TransactionIdTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/672f803dcd2b2fc651dac7cd9f39e08cef329fff.

deyanzz commented on sdk/tests/SerializationAndDeserializationTests.cc:
```diff
@@ -0,0 +1,199 @@
+  EXPECT_EQ(protoTransactionIdPtr->transactionvalidstart().seconds(), protoTimestampPtr->seconds());
+}
+
+TEST_F(SerializationAndDeserializationTests, DeserializeTransactionIdFromProtobufTest)
```
To write this test I used test [ProtobufTransactionId](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransactionIdTest.cc#L51) from [TransactionIdTest](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransactionIdTest.cc).

deyanzz replied:
Test `DeserializeTransactionIdFromProtobufTest` moved into `TransactionIdTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/672f803dcd2b2fc651dac7cd9f39e08cef329fff.

deyanzz commented on sdk/tests/SerializationAndDeserializationTests.cc:
```diff
@@ -0,0 +1,199 @@
+  EXPECT_EQ(transactionId.getValidTransactionTime(), now);
+}
+
+TEST_F(SerializationAndDeserializationTests, SerializeTransferToProtobufTest)
```
To write this test I used test [ProtoTransferTest](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransferTest.cc#L33) from [TransferTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransferTest.cc).

deyanzz replied:
Test `SerializeTransferToProtobufTest` into `TransferTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/abcec567906e347c29c9a94934ce4f14d47963d5.

deyanzz commented on sdk/tests/SerializationAndDeserializationTests.cc:
```diff
@@ -0,0 +1,199 @@
+  EXPECT_FALSE(protoAccountAmountPtr->is_approval());
+}
+
+TEST_F(SerializationAndDeserializationTests, DeserializeTransferFromProtobufTest)
```
To write this test I used test [ProtoTransferTest](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransferTest.cc#L33) from [TransferTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransferTest.cc).

deyanzz replied:
Test `DeserializeTransferToProtobufTest` into `TransferTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/abcec567906e347c29c9a94934ce4f14d47963d5.

deyanzz commented on sdk/tests/DeserializationTests.cc:
```diff
@@ -0,0 +1,254 @@
+  const std::shared_ptr<ECDSAPublicKey> mPublicKeyFromString = ECDSAPublicKey::fromString(mPublicKeyFromPrivate->toString());
+};
+
+TEST_F(DeserializationTests, DeserializeExchangeRateFromProtobufTest)
```
To write this test I used test [ProtobufExchangeRate](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/ExchangeRateTest.cc#L43) from [ExchangeRateTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/ExchangeRateTest.cc).

deyanzz replied:
Test `DeserializeExchangeRateFromProtobufTest` moved into `ExchangeRateTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/ce0137a5c63aec2af768426d458144e413528ca6.

deyanzz commented on sdk/tests/DeserializationTests.cc:
```diff
@@ -0,0 +1,254 @@
+            internal::TimestampConverter::fromProtobuf(*testProtoExchangeRateSecs).time_since_epoch().count());
+}
+
+TEST_F(DeserializationTests, DeserializeExchangeRateSetFromProtobufTest)
```
To write this test I used test [ProtobufExchangeRate](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/ExchangeRateTest.cc#L43) from [ExchangeRateTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/ExchangeRateTest.cc).

deyanzz replied:
Test `DeserializeExchangeRateSetFromProtobufTest ` moved into `ExchangeRateTest.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/a3879f92d8fa2ce3ce92699b7c572a84fbaca461.

deyanzz commented on sdk/tests/DeserializationTests.cc:
```diff
@@ -0,0 +1,254 @@
+  EXPECT_EQ(exchangeRateSet.getNextExchangeRate().value().getCurrentExchangeRate(), testCents * testCents / testHbar * testHbar);
+}
+
+TEST_F(DeserializationTests, DeserializeTransactionReceiptFromProtobufTest)
```
To write this test I used test [ProtobufTransactionReceipt](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransactionReceiptTest.cc#L46) from [TransactionReceiptTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransactionReceiptTest.cc).

deyanzz replied:
Removed test `DeserializeTransactionReceiptFromProtobufTest` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/8f2579891c1a4a154cdf7b967d8ad881cd0791cc.

deyanzz commented on sdk/tests/DeserializationTests.cc:
```diff
@@ -0,0 +1,254 @@
+            std::chrono::system_clock::time_point(std::chrono::seconds(secs)));
+}
+
+TEST_F(DeserializationTests, DeserializeTransactionRecordFromProtobufTest)
```
To write this test I used test [ProtobufTransactionRecord](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransactionRecordTest.cc#L34) from [TransactionRecordTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransactionRecordTest.cc).

deyanzz replied:
Remove test `DeserializeTransactionRecordFromProtobufTest ` from `DeserializationTests.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/679c14b787e9d21e506dfa2add686593a61f7863.

deyanzz commented on sdk/tests/DeserializationTests.cc:
```diff
@@ -0,0 +1,254 @@
+  EXPECT_EQ(txRecord.getTransferList().at(1).getAmount().toTinybars(), testTransferAmount);
+}
+
+TEST_F(DeserializationTests, DeserializeTransactionResponseFromProtobufTest)
```
To write this test I used test [ProtobufTransactionResponse](https://github.com/hashgraph/hedera-sdk-cpp/blob/c54139e8bc8f507e1f08d052b15eec3d188a0bd0/sdk/tests/TransactionResponseTest.cc#L31) from [TransactionResponseTest.cc](https://github.com/hashgraph/hedera-sdk-cpp/blob/main/sdk/tests/TransactionResponseTest.cc).

deyanzz replied:
Remove test `DeserializeTransactionResponseFromProtobufTest` from `DeserializationTests.cc` in commit https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/24e9524b0d8a82ce38fd17d6a04f8255a3d5e480.

rwalworth commented on sdk/tests/DeserializationTests.cc:
```diff
@@ -0,0 +1,261 @@
+  [[nodiscard]] inline const AccountId& getTestAccountId() const { return mAccountId; }
+  [[nodiscard]] inline const AccountId& getTestAccountIdFrom() const { return mAccountIdFrom; }
+  [[nodiscard]] inline const AccountId& getTestAccountIdTo() const { return mAccountIdTo; }
+  [[nodiscard]] inline const std::shared_ptr<ECDSAPublicKey>& getTestECDSAPublicKey() const { return mPublicKeyFromString; }
```
This doesn't look like it follows our formatting outlined in our `.clang-format` file. You should be able to set up your IDE to auto-format on save or on `git commit`.

rwalworth commented on sdk/tests/DeserializationTests.cc:
```diff
@@ -0,0 +1,261 @@
+  const uint64_t mContractNum = 3;
+  const int32_t cents = 2;
+  const int32_t hbar = 1;
+  const uint64_t seconds = 100ULL;
```
Can these variables follow our `m<NAME>` variable naming convention?

deyanzz replied:
Fixed in https://github.com/hashgraph/hedera-sdk-cpp/pull/176/commits/e201bc6d2c798a8aa3b1221be2d924bd8334ecda

rwalworth commented on sdk/tests/SerializationTests.cc:
```diff
@@ -0,0 +1,52 @@
+  // Then
+  EXPECT_NE(protobufECDSAPublicKey.get(), nullptr);
+  EXPECT_TRUE(protobufECDSAPublicKey.get()->has_ecdsa_secp256k1());
+}
```
You should test serializing a `EDD25519PrivateKey` as well.

deyanzz replied:
I double-checked it, but it seems `ED25519PrivateKey` doesn't have method `toProtobuf` and can not be serialized. Could you check again and tell me if I'm wrong?

rwalworth replied:
Sorry, I meant `ED25519PublicKey`

rwalworth reviewed (changes requested):
It seems like a lot of the tests that you have written here are more or less copies of tests that already existed. I don't see a point in replicating those tests; any additional testing that should happen with those objects should just be put into those tests that already exist. Also, I think it may be better to have tests involving a specific object to be contained in a test file for that object as opposed to dividing the suites up by function (e.g. serialization, deserialization, etc.).

deyanzz commented:
> It seems like a lot of the tests that you have written here are more or less copies of tests that already existed. I don't see a point in replicating those tests; any additional testing that should happen with those objects should just be put into those tests that already exist. Also, I think it may be better to have tests involving a specific object to be contained in a test file for that object as opposed to dividing the suites up by function (e.g. serialization, deserialization, etc.).

I was my fault I didn't communicate well that approach with you in the beginning. 
I removed the 3 files for separate testing (`SerializationAndDeserializationTests.cc`, `SerializationTests.cc` and `DeserializationTests.cc`) and moved all of the tests in the corresponding `*Test.cc` files. 
I agree it's better to stick tests involving a specific object to be contained in a test file for that object, so I did it already. Thanks for the advice!

rwalworth reviewed (approved):
LGTM
